core/types/subtyping.cc

```diff
@@ -1303,6 +1303,13 @@ bool Types::isSubTypeUnderConstraint(const GlobalState &gs, TypeConstraint &cons
                Types::isSubTypeUnderConstraint(gs, constr, t1, o2->right, mode);
     }
     if (a1 != nullptr) { // 4
+        auto *t2v = cast_type<TypeVar>(t2);
+        if (t2v && !constr.isSolved() && constr.hasLowerBound(t2v->sym) &&
+            ((const TypeConstraint &)constr).findLowerBound(t2v->sym) == Types::bottom()) {
+            constr.rememberIsSubtype(gs, t1, t2);
+            return true;
+        }
+
```

Comment:
This looks out of place. There is not similar code anywhere else in this function, so it surprises me to think that this is the solution.

i.e., most of the rest of this code is symmetric. that probably means we either need code like this everywhere, or the solution lies elsewhere.

Reply:
Is there a test we could write that gets a bug to manifest in the other cases? like

- t1 is a type var, and t2 is an or type
- t2 is a type var, and t1 is an or type
- ... etc

Reply:
I think "t2 is a type var and t1 is an or type" [already works](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20Parent%3B%20end%0Amodule%20Interface%3B%20end%0Aclass%20Child%20%3C%20Parent%0A%20%20include%20Interface%0Aend%0A%0Adef%20bar%0A%20%20x%20%3D%20T.let(%5B%5B%3Achild%2C%20Child.new%5D%5D%2C%20T%3A%3AArray%5B%5BSymbol%2C%20T.any(Parent%2C%20Interface)%5D%5D)%0A%20%20T.reveal_type(x.to_h)%0Aend).

If I understand the code correctly, "t1 is a type variable and t2 is an or type" should also fail to reveal the correct type, because we'd hit:

https://github.com/sorbet/sorbet/blob/c3344f71641f737916e1d38c8d64c83d225b3d5f/core/types/subtyping.cc#L1300-L1304

but we'd a) never check the right side of the or for subtyping and b) even if we did, we'd run into:

https://github.com/sorbet/sorbet/blob/c3344f71641f737916e1d38c8d64c83d225b3d5f/core/types/subtyping.cc#L990-L996
https://github.com/sorbet/sorbet/blob/c3344f71641f737916e1d38c8d64c83d225b3d5f/core/TypeConstraint.cc#L98-L122

I *think* the other combinations of type var and and/or type get the correct relations out of that code in `TypeConstraint`.

Reply:
Here is a test case that currently fails but I don't think it should:

```ruby
# typed: strict
extend T::Sig

sig do
  type_parameters(:U)
    .params(
      f: T.proc.params(x: T.type_parameter(:U)).void,
      x: T.type_parameter(:U)
    )
    .void
end
def foo(f, x)
end

f = T.let(
  proc {|x| nil},
  T.proc.params(x: T.any(String, Integer)).void
)

# These should be the constraints that we find:
#
#   T.type_parameter(:U) <: T.any(String, Integer)
#   Integer <: T.type_parameter(:U)
#
# which has a solution (really many solutions,
# of which TypeConstraint::solve should return any.)
#
# But in fact, these are the constraints that we find:
#
#   T.type_parameter(:U) <: String
#   Integer <: T.type_parameter(:U)
foo(f, 0)
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0Aextend%20T%3A%3ASig%0A%0Asig%20do%0A%20%20type_parameters(%3AU)%0A%20%20%20%20.params(%0A%20%20%20%20%20%20f%3A%20T.proc.params(x%3A%20T.type_parameter(%3AU)).void%2C%0A%20%20%20%20%20%20x%3A%20T.type_parameter(%3AU)%0A%20%20%20%20)%0A%20%20%20%20.void%0Aend%0Adef%20foo(f%2C%20x)%0Aend%0A%0Af%20%3D%20T.let(%0A%20%20proc%20%7B%7Cx%7C%20nil%7D%2C%0A%20%20T.proc.params(x%3A%20T.any(String%2C%20Integer)).void%0A)%0A%0A%23%20These%20should%20be%20the%20constraints%20that%20we%20find%3A%0A%23%0A%23%20%20%20T.type_parameter(%3AU)%20%3C%3A%20T.any(String%2C%20Integer)%0A%23%20%20%20Integer%20%3C%3A%20T.type_parameter(%3AU)%0A%23%0A%23%20which%20has%20a%20solution%20(really%20many%20solutions%2C%0A%23%20of%20which%20TypeConstraint%3A%3Asolve%20should%20return%20any.)%0A%23%0A%23%20But%20in%20fact%2C%20these%20are%20the%20constraints%20that%20we%20find%3A%0A%23%0A%23%20%20%20T.type_parameter(%3AU)%20%3C%3A%20String%0A%23%20%20%20Integer%20%3C%3A%20T.type_parameter(%3AU)%0Afoo(f%2C%200))

More generally, this is what I was getting at when I was saying this solution felt ad-hoc:

Almost every type system feature that deals with union types has a dual feature for intersection types.
It is a huge red flag when the code to handle one case does not share symmetry with the other case.
With a little extra effort, we can find test cases that go through the other branches that share the same bugs.

Reply:
That's fair.  The good news is that this test case is easily fixable by changing:

https://github.com/sorbet/sorbet/blob/c3344f71641f737916e1d38c8d64c83d225b3d5f/core/types/subtyping.cc#L1300-L1304

to not short-circuit.